consensus/hotstuff/forks/forks2.go

```diff
@@ -0,0 +1,557 @@
+//     validate blocks (incl. payload), i.e. QCs are processed as part of validated proposals.
+//
+// Possible error returns:
+//   - model.MissingBlockError if the parent does not exist in the forest (but is above
+//     the pruned view). From the perspective of Forks, this error is benign (no-op).
+//   - model.InvalidBlockError if the block is invalid (see `Forks.EnsureBlockIsValidExtension`
+//     for details). From the perspective of Forks, this error is benign (no-op).
+//   - model.ByzantineThresholdExceededError if conflicting QCs or conflicting finalized
+//     blocks have been detected (violating a foundational consensus guarantees). This
+//     indicates that there are 1/3+ Byzantine nodes (weighted by stake) in the network,
+//     breaking the safety guarantees of HotStuff (or there is a critical bug / data
+//     corruption). Forks cannot recover from this exception.
+//   - All other errors are potential symptoms of bugs or state corruption.
+func (f *Forks2) AddProposal(proposal *model.Block) error {
+	err := f.EnsureBlockIsValidExtension(proposal)
+	if err != nil {
+		return fmt.Errorf("validity check on block %v failed: %w", proposal.BlockID, err)
+	}
+	err = f.UnverifiedAddProposal(proposal)
+	if err != nil {
+		return fmt.Errorf("error storing block %v in Forks: %w", proposal.BlockID, err)
+	}
+	return nil
+}
+
+// UnverifiedAddCertifiedBlock appends the given certified block to the tree of pending
+// blocks and updates the latest finalized block (if applicable). Unless the parent is
+// below the pruning threshold (latest finalized view), we require that the parent is
+// already stored in Forks. Calling this method with previously processed blocks
+// leaves the consensus state invariant (though, it will potentially cause some
+// duplicate processing).
+// We assume that all blocks are fully verified. A valid block must satisfy all
+// consistency requirements; otherwise we have a bug in the compliance layer.
+// Notes:
+//   - UNVALIDATED: expects block to pass `Forks.EnsureBlockIsValidExtension(..)`
+//
+// Error returns:
+//   - model.ByzantineThresholdExceededError if conflicting QCs or conflicting finalized
+//     blocks have been detected (violating a foundational consensus guarantees). This
+//     indicates that there are 1/3+ Byzantine nodes (weighted by stake) in the network,
+//     breaking the safety guarantees of HotStuff (or there is a critical bug / data
+//     corruption). Forks cannot recover from this exception.
+//   - All other errors are potential symptoms of bugs or state corruption.
+func (f *Forks2) UnverifiedAddCertifiedBlock(certifiedBlock *model.CertifiedBlock) error {
+	if !f.IsProcessingNeeded(certifiedBlock.Block) {
+		return nil
+	}
+	err := f.store(certifiedBlock.Block)
+	if err != nil {
+		return fmt.Errorf("storing block %v in Forks failed %w", certifiedBlock.Block.BlockID, err)
+	}
+	err = f.checkForConflictingQCs(certifiedBlock.CertifyingQC)
```

Comment:
can we move this check so we execute it before `store`?

Reply:
👉 https://github.com/onflow/flow-go/pull/4190/commits/2231c4a0b0d3c2fec4a442b088b68e5142216ec8
(required some moderate restructuring of logic to avoid duplicating code or different behaviours of `AddProposal` vs `AddCertifiedBlock`)